### PR TITLE
[Testing] Add basic setup and sample tests for Codeception System Tests

### PR DESCRIPTION
**Pull request by yvesh**
### Summary of Changes

This pull request adds the basic setup and configuration for the Joomla browser based System Tests. It contains sample tests for com_users. During the next weeks more tests are going to be ported from the [system tests repository
](https://github.com/joomla-projects/gsoc16_browser-automated-tests). Additionally Continuous Integration will be activated running the test suite for every PR and commit.

### Testing Instructions

For a complete description check the README.md in tests/codeception

Prerequisites:

* Webserver with PHP installed
* Google Chrome
* Composer
* PHP 5.6 Cli (or later)
* Java JDK (8 or later)

Step by Step:

* Navigate in your terminal to `tests/codeception`
* Do a `composer install`
* cp and modify the `acceptance.suite.dist.yml` to `acceptance.suite.yml`
* Navigate back to the root folder of Joomla
* Run `tests/codeception/vendor/bin/robo run:tests`

### Expected result

Joomla has System tests

### Actual result

Joomla has no System tests

### Documentation Changes Required

Yes, parts are already integrated and will be updated the next days.

Thank you to all people working on this topic in the last years!

### CODE REVIEW

File: tests/codeception/README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+Test Automation demands considerable investments of money and resources. Successive development cycles will require execution of same test suite repeatedly. Using a test automation tool it's possible to record this test suite and re-play it as required. Once the test suite is automated, no human intervention is required.
+
+### Installation
+
+1. Clone this repository
+
+2. Install `composer` in your system. Read more about [how to install composer](https://getcomposer.org/doc/00-intro.md) here.
+
+3. Install composer packages using following steps from root directory of this project.
```
Comment (brianteeman) [inline tests/codeception/README.md]: using the following

File: tests/codeception/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+We are using `composer.json` file for `tests/codeception` folder, so that you will have to run composer install from tests directory.
```
Comment (brianteeman) [inline tests/codeception/README.md]: from the tests directory

File: tests/codeception/README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+```bash
+$ tests/codeception/vendor/bin/robo run:test
+```
+
+**_Windows_**
+```cmd
+$ tests\codeception\vendor\bin\robo run:test
+```
+
+If you want to see steps then you can use `--steps` option of codeception. Check [full codecept command list here](http://codeception.com/docs/reference/Commands#Run)_
```
Comment (brianteeman) [inline tests/codeception/README.md]: see the steps

File: tests/codeception/README.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+2. Install `composer` in your system. Read more about [how to install composer](https://getcomposer.org/doc/00-intro.md) here.
+
+3. Install composer packages using following steps from root directory of this project.
+We are using `composer.json` file for `tests/codeception` folder, so that you will have to run composer install from tests directory.
+
+    ```bash
+    $ cd tests/codeception && composer install
+    ```
+
+4. Copy `tests/codeception/acceptance.suite.dist.yml` to `tests/codeception/acceptance.suite.yml` and change settings according to your webserver.
```
Comment (brianteeman) [inline tests/codeception/README.md]: change the settings

File: tests/codeception/README.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+```
+
+**_Windows_**
+```cmd
+$ tests\codeception\vendor\bin\robo run:test
+```
+
+If you want to see steps then you can use `--steps` option of codeception. Check [full codecept command list here](http://codeception.com/docs/reference/Commands#Run)_
+
+**Note**:You can modify the timeout time by setting the value of **TIMEOUT** constant lower for fast machines and higher for slow computers. The constant located in the file `tests/codeception/acceptance/_bootstrap.php`
```
Comment (brianteeman) [inline tests/codeception/README.md]: +**Note**: You can modify the timeout time by setting the value of  the **TIMEOUT** constant lower for fast machines and higher for slow computers.

File: tests/codeception/_support/AcceptanceTester.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
+<?php
+/**
+ * @package     Joomla.Test
+ * @subpackage  AcceptanceTester
+ *
+ * @copyright   Copyright (C) 2005 - 2016 Open Source Matters, Inc. All rights reserved.
```
Comment (brianteeman) [inline tests/codeception/_support/AcceptanceTester.php]: 2017

File: tests/codeception/_support/AcceptanceTester.php [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+ * @license     GNU General Public License version 2 or later; see LICENSE
```
Comment (brianteeman) [inline tests/codeception/_support/AcceptanceTester.php]: LICENSE.txt

File: tests/codeception/_support/FunctionalTester.php [hunk 8]
```diff
@@ -0,0 +1,37 @@
+<?php
+/**
+ * @package     Joomla.Test
+ * @subpackage  FunctionalTester
+ *
+ * @copyright   Copyright (C) 2005 - 2016 Open Source Matters, Inc. All rights reserved.
```
Comment (brianteeman) [inline tests/codeception/_support/FunctionalTester.php]: 2017
LICENSE.txt

File: tests/codeception/_support/AcceptanceTester.php (hunk 7 above)
Comment (brianteeman) [inline tests/codeception/_support/AcceptanceTester.php]: and the same for all new files

File: tests/codeception/_support/Helper/JoomlaDb.php [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+	 */
+	public function haveInDatabase($table, array $data)
+	{
+		$table = $this->addPrefix($table);
+
+		return parent::haveInDatabase($table, $data);
+	}
+
+	/**
+	 * See an entry in the database
```
Comment (brianteeman) [inline tests/codeception/_support/Helper/JoomlaDb.php]: Would  this be better as Find in the database and not found in the database?

File: tests/codeception/acceptance/administrator/components/com_users/UserCest.php [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+		$I->click(Administrator\UserManagerPage::$saveButton);
+		$I->waitForText(Administrator\UserManagerPage::$pageTitleText);
+		$I->see(Administrator\UserManagerPage::$successMessage, Administrator\AdminPage::$systemMessageContainer);
+
+		$I->checkForPhpNoticesOrWarnings();
+	}
+
+	/**
+	 * Edit an user
```
Comment (brianteeman) [inline tests/codeception/acceptance/administrator/components/com_users/UserCest.php]: Edit a user


(the rule is that it is an when before a vowel EXCEPT when the vowel is pronounced as a consonant - which it is in this case YOU-SIR

File: tests/codeception/functional.suite.yml [hunk 11]
```diff
@@ -0,0 +1,11 @@
+# Codeception Test Suite Configuration
+#
+# Suite for functional (integration) tests
+# Emulate web requests and make application process them
+# Include one of framework modules (Symfony2, Yii2, Laravel5) to use it
```
Comment (brianteeman) [inline tests/codeception/functional.suite.yml]: one of the framework

Comment (yvesh) [inline tests/codeception/functional.suite.yml]: This file was generated by codeception, but i still fixed it for us :) Thank you @brianteeman !!

**Comment by zero-24** [comment]
@yvesh did you already run the script for the header that I did not updated yesterday?

**Comment by yvesh** [comment]
@zero-24 Updated it manually now, because we have too many different subpackage names.

**Comment by SniperSister** [comment]
I have tested this item :white_check_mark: successfully on 6ddf0c253c920019c1b7bbcd69235beceedb91a5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16743">issues.joomla.org/tracker/joomla-cms/16743</a>.</sub>

File: tests/codeception/README.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+```bash
+$ tests/codeception/vendor/bin/robo run:test
+```
+
+**_Windows_**
+```cmd
+$ tests\codeception\vendor\bin\robo run:test
+```
+
+If you want to see the steps then you can use `--steps` option of codeception. Check [full codecept command list here](http://codeception.com/docs/reference/Commands#Run)_
```
Comment (brianteeman) [inline tests/codeception/README.md]: remove trailing _

**Comment by dneukirchen** [comment]
I have tested this item ✅ successfully

bildschirmfoto 2017-06-18 um 12 58 57

**Pull request merged**